Feature/provisioner apis

albertshau opened the pull request.

albertshau committed:
adding apis and store for provisioner information, such as the
number of workers running on each provisioner for each tenant.

albertshau committed:
removing unused code with provisioner status and assign time, and
adding a convenience function for getting a db connection with
auto commit on since thats the vast majority.

albertshau committed:
fixing heartbeat and expire bugs, and adding more logging.

albertshau committed:
adding more unit tests.

albertshau committed:
add documentation for superadmin provisioner apis and changelog.

albertshau committed:
adding docs for config settings for provisioner requests and
doing some cleanup to simplify building the http client.

albertshau committed:
Merge remote-tracking branch 'origin/develop' into feature/provisioner-apis

Conflicts:
	server/src/main/java/com/continuuity/loom/common/conf/Constants.java
	server/src/main/java/com/continuuity/loom/common/queue/guice/QueueModule.java
	server/src/main/java/com/continuuity/loom/scheduler/JobScheduler.java
	server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java
	server/src/test/java/com/continuuity/loom/BaseTest.java
	server/src/test/java/com/continuuity/loom/http/LoomServiceTestBase.java

poornachandra commented on server/src/main/java/com/continuuity/loom/common/queue/QueueType.java:
```diff
@@ -24,7 +24,8 @@
   CALLBACK("/clustermanager/callback"),
-  PROVISIONER("/clustermanager/nodeprovision");
+  PROVISIONER("/clustermanager/nodeprovision"),
+  BALANCER("/clustermanager/balancer");
```
Now that we have so many queues, I think it would be good to have a brief description of each queue.

albertshau replied:
done

poornachandra commented on server/src/main/java/com/continuuity/loom/store/DBQueryHelper.java:
```diff
@@ -143,6 +143,26 @@ public static void dropDerbyDB() {
+   * Queries for a single number, returning the value of the number.
+   *
+   * @param statement PreparedStatement of the query, ready for execution.
+   * @return Result of the query.
```
@return Result of the query, or 0 if no results.

albertshau replied:
done

poornachandra commented on server/src/main/java/com/continuuity/loom/store/provisioner/ProvisionerStore.java:
```diff
@@ -0,0 +1,35 @@
+ */
+public interface ProvisionerStore extends Service {
+
+  Collection<Provisioner> getAllProvisioners() throws IOException;
```
Missing javadocs for interface methods.

albertshau replied:
done

poornachandra commented on server/src/main/java/com/continuuity/loom/store/provisioner/SQLProvisionerStore.java:
```diff
@@ -0,0 +1,533 @@
+ * two tables, one for storing provisioner information and another for maintaining an index from tenants to
+ * provisioners.
+ */
+public class SQLProvisionerStore extends AbstractScheduledService implements ProvisionerStore {
```
Does SQLProvisionerStore need to be a AbstractScheduledService?

albertshau replied:
good catch, no it does not. Was part of an incomplete implementation that I didn't remove.

poornachandra commented on docs/docs/source/rest/provisioners.rst:
```diff
@@ -0,0 +1,152 @@
+        http://<loom-server>:<superadmin-port>/<version>/provisioners/p1
+ $ {
+       "id": "p1",
+       "host": "localhost",
```
localhost is probably not a good example.

albertshau replied:
changed to "host1"

poornachandra commented on server/src/main/java/com/continuuity/loom/http/handler/LoomProvisionerHandler.java:
```diff
@@ -0,0 +1,164 @@
+import java.io.Reader;
+
+/**
+ * Handles requests to take and finish tasks from provisioners.
```
The javadoc does not match the APIs

albertshau replied:
done

poornachandra commented on server/src/main/java/com/continuuity/loom/provisioner/Provisioner.java:
```diff
@@ -0,0 +1,241 @@
+    this.port = port;
+    this.capacityTotal = capacityTotal;
+    this.usage = usage == null ? ImmutableMap.<String, Integer>of() : ImmutableMap.copyOf(usage);
+    this.assignments = assignments == null ? Maps.<String, Integer>newHashMap() : Maps.newHashMap(assignments);
```
It is probably better to use concurrent map here.

albertshau replied:
done

poornachandra commented on server/src/main/java/com/continuuity/loom/provisioner/TenantProvisionerService.java:
```diff
@@ -0,0 +1,289 @@
+    this.provisionerStore = provisionerStore;
+    this.tenantStore = tenantStore;
+    this.provisionerRequestService = provisionerRequestService;
+    this.lock = new ZKInterProcessReentrantLock(zkClient, Constants.TENANT_NAMESPACE);
```
Can you add a brief description of why we use a single lock for all tenants?

albertshau replied:
done

poornachandra commented on server/src/main/java/com/continuuity/loom/provisioner/TenantProvisionerService.java:
```diff
@@ -0,0 +1,289 @@
+      // send requests to provisioners to delete tenant
+      // if the db dies or something in between, the client making the request will get a 500.  No way to rollback,
+      // so client would have to resend the delete when things are ok again.
+      // TODO: what happens if we crash here? workaround would be to restart provisioners...
```
If the delete request goes on a queue then it will get retried when the server comes back up, right?

albertshau replied:
yeah though we'd need to put in the logic to retry queue elements and would have to make sure all the provisioners for the tenant get placed into a single queue element.  Will put that in a todo to be handled later.

albertshau replied:
hm there are still race conditions since 2 tables are being updated plus the provisioner calls.  Another option would be to split this into 2 calls and only allow deleting a tenant if its workers are already set to 0.

poornachandra replied:
Right.

albertshau replied:
going to go with the 2nd option of forcing workers to be set to 0 before deleting.  Simplifies the logic and also makes it explicit to the caller that workers are stopped and no additional tasks will be kicked off when they delete the tenant.

poornachandra commented on server/src/main/java/com/continuuity/loom/provisioner/TenantProvisionerService.java:
```diff
@@ -0,0 +1,289 @@
+    try {
+      // if a provisioner died earlier leaving us under capacity, and this provisioner adds capacity, we want to
+      // rebalance the workers after writing the provisioner.
+      boolean rebalanceWorkers = provisionerStore.getFreeCapacity() == 0;
```
Should we just re-balance every time?

albertshau replied:
agreed, this would be needed anyway if we have different balancing policies.

poornachandra commented on server/src/main/java/com/continuuity/loom/provisioner/TenantProvisionerService.java:
```diff
@@ -0,0 +1,289 @@
+  // Currently a greedy approach, just add to first available.
+  private void addWorkers(String tenantId, int numToAdd) throws CapacityException, IOException {
+    for (Provisioner provisioner : provisionerStore.getProvisionersWithFreeCapacity()) {
+      int numAdded = provisioner.tryAddTenantAssignments(tenantId, numToAdd);
```
You can add one more check - if numToAdd <= 0 then break;

albertshau replied:
good point, done.

poornachandra commented on server/src/main/java/com/continuuity/loom/provisioner/TenantProvisionerService.java:
```diff
@@ -0,0 +1,289 @@
+          // TODO: what if this fails?
+          LOG.error("Could not write tenant {} to provisioner {}. The provisioner appears broken, deleting it and " +
+                     "rebalancing its tenant workers", tenantId, provisioner.getId());
+          deleteProvisioner(provisioner);
```
Can you add a comment on whether a rebalance will be triggered  during this call or after this call is complete.

albertshau replied:
done

poornachandra commented on server/src/main/java/com/continuuity/loom/provisioner/guice/ProvisionerModule.java:
```diff
@@ -0,0 +1,17 @@
+import com.google.inject.Scopes;
+
+/**
+ *
```
Javadoc missing.

albertshau replied:
done

albertshau committed:
pull request changes.  Adding more comments and javadocs. Also,
make sure a tenants workers have been set to 0 before deleting it.
This is to make it clear to the caller the behavior of the delete.
No additional tasks will be performed for the tenant since it has
no workers.

albertshau committed:
Merge remote-tracking branch 'origin/develop' into feature/provisioner-apis

poornachandra commented on server/src/main/java/com/continuuity/loom/scheduler/Scheduler.java:
```diff
@@ -154,5 +168,6 @@ private void unschedule() {
       }
     }
     scheduledFutures.clear();
+    tenantProvisionerService.stopAndWait();
```
Once a Guava service is stopped, it cannot be restarted again in schedule. Use a provider or something to get a new object the next time when it becomes leader.

poornachandra replied:
Or else, make changes so that the service can keep running when server becomes follower.

albertshau replied:
good catch.  Will move the provisioner timeout logic to a separate class so the service can always be running.

poornachandra commented on server/src/main/java/com/continuuity/loom/store/provisioner/SQLProvisionerStore.java:
```diff
@@ -0,0 +1,533 @@
+  }
+
+  @Override
+  public Collection<Provisioner> getIdleProvisioners(long idleTimestamp) throws IOException {
```
Idle does not convey the meaning of this method, right?

albertshau replied:
good point.  Will rename to something along the lines of 'unresponsive'

poornachandra commented on server/src/main/java/com/continuuity/loom/provisioner/TenantProvisionerService.java:
```diff
@@ -0,0 +1,289 @@
+
+  public void handleHeartbeat(String provisionerId, ProvisionerHeartbeat heartbeat)
+    throws IOException, MissingEntityException {
+    lock.acquire();
```
Is this lock required?

albertshau replied:
good point, it is not required. will remove and add a comment.

poornachandra commented on server/src/main/java/com/continuuity/loom/store/provisioner/SQLProvisionerStore.java:
```diff
@@ -0,0 +1,533 @@
+        statement.close();
+      }
+    } finally {
+      statement.close();
```
I think there is an extra statement.close()

albertshau replied:
yeah there is an extra try finally. will fix it.

poornachandra commented on docs/docs/source/rest/tenants.rst:
```diff
@@ -156,8 +156,8 @@ HTTP Responses
      - Description
    * - 200 (OK)
      - If delete was successful
-   * - 404 (NOT FOUND)
```
What happens when you don't find a tenant?

albertshau replied:
this was incorrect documentation to begin with.  Deleting a nonexistant entity returns a 200 so that it's idempotent.

poornachandra commented:
Few more comments. You can merge after that is done :+1:

albertshau committed:
more pull request changes. Removing unused code, adding more
comments and moving provisioner timeout logic to its own thread
so that the TenantProvisionerService can be used everywhere
without starting/stopping.

albertshau committed:
Merge remote-tracking branch 'origin/develop' into feature/provisioner-apis

Conflicts:
	CHANGELOG.md
	server/src/main/java/com/continuuity/loom/codec/json/JsonSerde.java
	server/src/main/java/com/continuuity/loom/http/handler/LoomSuperadminHandler.java
	server/src/main/java/com/continuuity/loom/runtime/LoomServerMain.java
	server/src/main/java/com/continuuity/loom/store/entity/SQLAdminEntityStoreView.java
	server/src/test/java/com/continuuity/loom/BaseTest.java
	server/src/test/java/com/continuuity/loom/http/LoomSuperadminHandlerTest.java
